crates/typst-syntax/src/parser.rs

```diff
@@ -1494,23 +1497,70 @@ fn pattern_leaf<'s>(
+}
+
+/// The starting point of any trivia (whitespace/comments) preceding the current token.
+#[derive(Clone)]
+struct TriviaStart {
+    /// Number of preceding trivia `nodes` (for avoiding trivia when wrapping).
+    num: usize,
+    /// Offset into `text` of the trivia start (for moving the lexer's cursor back).
+    offset: usize,
+}
+
+/// Manages parsing of a stream of tokens into a tree of SyntaxNodes.
+///
+/// This acts like a bottom-up (shift-reduce) parser, where we investigate a current token
+/// and either:
+/// 1. Eat (shift) another token, putting `current` into the `nodes` vector.
+/// 2. Wrap (reduce) the nodes up to a marker (excluding `current`) into a corresponding
+///    SyntaxKind.
+///
+/// In the end we have a nested tree of SyntaxNodes as a Concrete Syntax Tree. This will
+/// have all irrelevant tokens (whitespace, comments, code parens, commas in function
+/// args, etc) removed when interpreted as an Abstract Syntax Tree for evaluation. The
+/// Concrete Syntax Tree is important for IDE features and highlighting.
 struct Parser<'s> {
+    /// The entire source text, shared with the lexer, usually accessed via usize offset.
     text: &'s str,
+    /// A lexer over the source text with multiple possible modes. Our token generator.
     lexer: Lexer<'s>,
-    prev_end: usize,
-    current_start: usize,
-    current: SyntaxKind,
+    /// The current token being evaluated, not yet present in `nodes`. This acts like a
+    /// single item of lookahead for the parser.
+    /// Invariant: This should never be trivia in Math/Code mode.
+    current: Token<'s>,
+    /// The start of any trivia before `token` in Math/Code mode. Markup parses trivia
+    /// manually and doesn't use this.
+    prev_trivia: Option<TriviaStart>,
+    /// Whether the parser is balanced over open/close delimiters when not actively
+    /// matching delimiters. This only ever transitions from true to false.
     balanced: bool,
+    /// Nodes representing the structural syntax tree of anything previously parsed. Does
+    /// include the nodes from `prev_trivia`, but does not include `current`
     nodes: Vec<SyntaxNode>,
+    /// The stack of our lexer modes, these are pushed and popped by the `enter()` and
+    /// `exit()` functions when we switch modes.
     modes: Vec<LexMode>,
+    /// A stack of newline modes used by Code Mode to determine when an expression should
+    /// end or continue in the presence of newlines.
     newline_modes: Vec<NewlineMode>,
+    /// Stored parser checkpoints at a given text index for efficient math paren-expr
+    /// parsing, see comments in `expr_with_paren` (also known as packrat parsing).
     memo: HashMap<usize, (Range<usize>, Checkpoint<'s>)>,
+    /// The stored parse results at each checkpoint.
     memo_arena: Vec<SyntaxNode>,
 }
 
 /// How to proceed with parsing when seeing a newline.
-#[derive(Clone)]
+///
+/// TODO idea: Make LexMode::Math contain the newline mode internally, then we can remove
```

Comment:
do you mean `LexMode::Code`? Perhaps, not sure whether it would work, but feel free to try.

Reply:
Yep lol.

I ended up trying this and I think it worked out well. I'm pushing this as a new commit, see below.